[Error] Gates issue

### **Steps to reproduce:**

1) Go to the Lorencia -> Devias gate
2) Enter Devias
3) Go back to Lorencia

### **Result:**

**No error occure but server and client stuck.**
Short look: stuck occures at the moment when
```
 public void EnterGate(Player player, EnterGate enterGate){
...
if (player.CurrentMap.Definition.EnterGates.Contains(enterGate) && !(player.X >= enterGate.X1 - INACCURACY &&
                  player.X <= enterGate.X2 + INACCURACY &&
                  player.Y >= enterGate.Y1 - INACCURACY &&
                  player.Y <= enterGate.Y2 + INACCURACY))
            {
                return;
            }
...
}
```
causes _return_. Return triggers in 90% of cases due to small INACCURACY (currently its **2**). _Tested with **4** - everything is fine._

After INACCURACY changed to 4 and it become working next error occured:
```
[ERROR] 2018-07-14 15:18:37,227 [MUnique.OpenMU.GameServer.RemoteView.RemotePlayer] [0] - System.NullReferenceException: Ссылка на объект не указывает на экземпляр объекта.
   в MUnique.OpenMU.GameLogic.Player.WarpTo(ExitGate gate) в D:\MuProject\TheServer\OpenMU-1\src\GameLogic\Player.cs:строка 502
   в MUnique.OpenMU.GameLogic.PlayerActions.WarpGateAction.EnterGate(Player player, EnterGate enterGate) в D:\MuProject\TheServer\OpenMU-1\src\GameLogic\PlayerActions\WarpGateAction.cs:строка 55
   в MUnique.OpenMU.GameServer.MessageHandler.WarpGateHandler.HandlePacket(Player player, Byte[] packet) в D:\MuProject\TheServer\OpenMU-1\src\GameServer\MessageHandler\WarpGateHandler.cs:строка 50
   в MUnique.OpenMU.GameServer.ConfigurablePacketHandler.HandlePacket(Player player, Byte[] packet) в D:\MuProject\TheServer\OpenMU-1\src\GameServer\ConfigurablePacketHandler.cs:строка 56
   в MUnique.OpenMU.GameServer.RemoteView.RemotePlayer.PacketReceived(Byte[] buffer) в D:\MuProject\TheServer\OpenMU-1\src\GameServer\RemoteView\RemotePlayer.cs:строка 71
```

### **Steps to reproduce:**

1) Change INACCURACY  to 4
2) Enter Devias via Lorencia gate
3) Leave Devias -> Enter Lorencia
4) Enter Devias again
5) Try to leave Devias via Lorencia gate

**Upd.**
Error occures at 

```
public void EnterGate(Player player, EnterGate enterGate){
...
player.WarpTo(enterGate.TargetGate);
...
}
```
Because of enterGate.TargetGate == NULL. 
targetGateId is ok but rawTargetGate and targetGate become NULL.

--- a/src/GameLogic/PlayerActions/WarpGateAction.cs
+++ b/src/GameLogic/PlayerActions/WarpGateAction.cs
@@ -44,10 +44,10 @@ namespace MUnique.OpenMU.GameLogic.PlayerActions
                 return;
             }
 
-            if (player.CurrentMap.Definition.EnterGates.Contains(enterGate) && !(player.X >= enterGate.X1 - INACCURACY &&
-                  player.X <= enterGate.X2 + INACCURACY &&
-                  player.Y >= enterGate.Y1 - INACCURACY &&
-                  player.Y <= enterGate.Y2 + INACCURACY))
+            if (player.CurrentMap.Definition.EnterGates.Contains(enterGate) && !(player.WalkTarget.X >= enterGate.X1 - INACCURACY &&
+                  player.WalkTarget.X <= enterGate.X2 + INACCURACY &&
+                  player.WalkTarget.Y >= enterGate.Y1 - INACCURACY &&
+                  player.WalkTarget.Y <= enterGate.Y2 + INACCURACY))
             {
                 return;
             }

--- a/src/Persistence/EntityFramework/Json/DelayedReferenceResolvingConverter.cs
+++ b/src/Persistence/EntityFramework/Json/DelayedReferenceResolvingConverter.cs
@@ -1,0 +1,113 @@
+﻿// <copyright file="DelayedReferenceResolvingConverter.cs" company="MUnique">
+// Licensed under the MIT License. See LICENSE file in the project root for full license information.
+// </copyright>
+
+namespace MUnique.OpenMU.Persistence.EntityFramework.Json
+{
+    using System;
+    using Newtonsoft.Json;
+    using Newtonsoft.Json.Linq;
+
+    /// <summary>
+    /// A json converter which is able to delay calls to the reference resolver when the resolver is unable to resolve the reference.
+    /// </summary>
+    /// <seealso cref="Newtonsoft.Json.JsonConverter" />
+    public class DelayedReferenceResolvingConverter : JsonConverter
+    {
+        /// <summary>
+        /// The object which is currently getting populated by the JsonSerializer.
+        /// </summary>
+        private object currentlyPopulating;
+
+        private Action delayedReferenceResolveActions;
+
+        /// <inheritdoc />
+        public override bool CanWrite => false;
+
+        /// <inheritdoc />
+        public override bool CanRead => true;
+
+        /// <inheritdoc />
+        public override bool CanConvert(Type objectType) => !objectType.IsEnum && !objectType.IsGenericType && (objectType.Namespace?.StartsWith("MUnique.OpenMU.") ?? false);
+
+        /// <summary>
+        /// Resolves the delayed references.
+        /// </summary>
+        public void ResolveDelayedReferences() => this.delayedReferenceResolveActions?.Invoke();
+
+        /// <inheritdoc/>
+        public override void WriteJson(JsonWriter writer, object value, JsonSerializer serializer)
+        {
+            throw new InvalidOperationException("Use default serialization.");
+        }
+
+        /// <inheritdoc/>
+        public override object ReadJson(JsonReader reader, Type objectType, object existingValue, JsonSerializer serializer)
+        {
+            if (reader.TokenType == JsonToken.Null)
+            {
+                return null;
+            }
+
+            var jsonObject = JObject.Load(reader);
+            if (jsonObject == null)
+            {
+                return null;
+            }
+
+            if (jsonObject["$ref"] != null)
+            {
+                return this.ResolveObjectReference(reader, serializer, jsonObject);
+            }
+
+            return this.CreateObject(objectType, serializer, jsonObject);
+        }
+
+        private object ResolveObjectReference(JsonReader reader, JsonSerializer serializer, JObject jsonObject)
+        {
+            var id = jsonObject["$ref"].Value<string>();
+            if (string.IsNullOrEmpty(id))
+            {
+                return null;
+            }
+
+            var resolvedObject = serializer.ReferenceResolver.ResolveReference(serializer.Context, id);
+            if (resolvedObject == null)
+            {
+                var property = this.currentlyPopulating.GetType().GetProperty(reader.Path);
+                if (property != null)
+                {
+                    var currentParent = this.currentlyPopulating;
+                    this.delayedReferenceResolveActions += () =>
+                    {
+                        var resolved = serializer.ReferenceResolver.ResolveReference(serializer.Context, id);
+                        property.SetValue(currentParent, resolved);
+                    };
+                }
+            }
+
+            return resolvedObject;
+        }
+
+        private object CreateObject(Type objectType, JsonSerializer serializer, JObject jsonObject)
+        {
+            // var createdObject = jsonObject.ToObject(objectType);
+            var createdObject = Activator.CreateInstance(objectType);
+
+            // We need to remember the currently populating object for the next properties on the same hierarchy.
+            // We could use a Stack<object> here, as well - ofc, parentPopulating lives on the "stack" too ;)
+            var parentPopulating = this.currentlyPopulating;
+            try
+            {
+                this.currentlyPopulating = createdObject;
+                serializer.Populate(jsonObject.CreateReader(), createdObject);
+            }
+            finally
+            {
+                this.currentlyPopulating = parentPopulating;
+            }
+
+            return createdObject;
+        }
+    }
+}

--- a/src/Persistence/EntityFramework/Json/JsonObjectDeserializer.cs
+++ b/src/Persistence/EntityFramework/Json/JsonObjectDeserializer.cs
@@ -33,25 +33,19 @@ namespace MUnique.OpenMU.Persistence.EntityFramework.Json
             serializer.ReferenceResolver = referenceResolver;
             serializer.Converters.Add(new BinaryAsHexJsonConverter());
 
+            DelayedReferenceResolvingConverter deferredConverter = null;
             if (this.AreCircularReferencesExpected)
             {
-                // This is a very naive approach to fix references to objects which occur later in the document.
-                // The first run just have the task to fills the references in the IdReferenceResolver.
-                // Unfortunately there is currently no better way, because we have circular references.
-                var jsonString = textReader.ReadToEnd();
-                using (var stringReader = new StringReader(jsonString))
-                using (var jsonReader = new JsonTextReader(stringReader))
-                {
-                    serializer.Deserialize<T>(jsonReader);
-                }
-
-                textReader = new StringReader(jsonString);
+                // For circular references, we add a converter which collects actions to resolve unresolved references, so they can be resolved after deserializing.
+                deferredConverter = new DelayedReferenceResolvingConverter();
+                serializer.Converters.Add(deferredConverter);
             }
 
             using (textReader)
             using (var jsonReader = new JsonTextReader(textReader))
             {
                 var result = serializer.Deserialize<T>(jsonReader);
+                deferredConverter?.ResolveDelayedReferences();
                 return result;
             }
         }

--- a/src/Persistence/EntityFramework/MUnique.OpenMU.Persistence.EntityFramework.csproj
+++ b/src/Persistence/EntityFramework/MUnique.OpenMU.Persistence.EntityFramework.csproj
@@ -203,6 +203,7 @@
     <Compile Include="Json\AccountJsonObjectLoader.cs" />
     <Compile Include="Json\AccountJsonQueryBuilder.cs" />
     <Compile Include="Json\BinaryAsHexJsonConverter.cs" />
+    <Compile Include="Json\DelayedReferenceResolvingConverter.cs" />
     <Compile Include="Json\GameConfigurationJsonObjectLoader.cs" />
     <Compile Include="Json\GameConfigurationJsonQueryBuilder.cs" />
     <Compile Include="Json\MultipleSourceReferenceResolver.cs" />

--- a/src/Persistence/Initialization/Items/Armors.cs
+++ b/src/Persistence/Initialization/Items/Armors.cs
@@ -361,7 +361,7 @@ namespace MUnique.OpenMU.Persistence.Initialization.Items
             defenseBonus.PowerUpDefinition.Boost = this.Context.CreateNew<PowerUpDefinitionValue>();
             defenseBonus.PowerUpDefinition.Boost.ConstantValue.AggregateType = AggregateType.Multiplicate;
             defenseBonus.PowerUpDefinition.Boost.ConstantValue.Value = bonus;
-            defenseBonus.PowerUpDefinition.TargetAttribute = Stats.DefenseBase;
+            defenseBonus.PowerUpDefinition.TargetAttribute = Stats.DefenseBase.GetPersistent(this.GameConfiguration);
             return defenseBonus;
         }
 
@@ -390,7 +390,7 @@ namespace MUnique.OpenMU.Persistence.Initialization.Items
             defenseRateBonus.PowerUpDefinition.Boost = this.Context.CreateNew<PowerUpDefinitionValue>();
             defenseRateBonus.PowerUpDefinition.Boost.ConstantValue.AggregateType = AggregateType.Multiplicate;
             defenseRateBonus.PowerUpDefinition.Boost.ConstantValue.Value = 1.1f;
-            defenseRateBonus.PowerUpDefinition.TargetAttribute = Stats.DefenseRatePvm;
+            defenseRateBonus.PowerUpDefinition.TargetAttribute = Stats.DefenseRatePvm.GetPersistent(this.GameConfiguration);
 
             var defenseBonus = new Dictionary<int, ItemOption>
             {
@@ -433,7 +433,7 @@ namespace MUnique.OpenMU.Persistence.Initialization.Items
             if (defense > 0)
             {
                 var powerUp = this.Context.CreateNew<ItemBasePowerUpDefinition>();
-                powerUp.TargetAttribute = Stats.DefenseBase;
+                powerUp.TargetAttribute = Stats.DefenseBase.GetPersistent(this.GameConfiguration);
                 powerUp.BaseValue = defense;
                 this.shieldDefenseBonusPerLevel.ForEach(powerUp.BonusPerLevel.Add);
                 shield.BasePowerUpAttributes.Add(powerUp);
@@ -442,7 +442,7 @@ namespace MUnique.OpenMU.Persistence.Initialization.Items
             if (defenseRate > 0)
             {
                 var powerUp = this.Context.CreateNew<ItemBasePowerUpDefinition>();
-                powerUp.TargetAttribute = Stats.DefenseRatePvm;
+                powerUp.TargetAttribute = Stats.DefenseRatePvm.GetPersistent(this.GameConfiguration);
                 powerUp.BaseValue = defenseRate;
                 this.defenseBonusPerLevel.ForEach(powerUp.BonusPerLevel.Add);
                 shield.BasePowerUpAttributes.Add(powerUp);
@@ -472,7 +472,7 @@ namespace MUnique.OpenMU.Persistence.Initialization.Items
             if (defense > 0)
             {
                 var powerUp = this.Context.CreateNew<ItemBasePowerUpDefinition>();
-                powerUp.TargetAttribute = Stats.DefenseBase;
+                powerUp.TargetAttribute = Stats.DefenseBase.GetPersistent(this.GameConfiguration);
                 powerUp.BaseValue = defense;
                 this.defenseBonusPerLevel.ForEach(powerUp.BonusPerLevel.Add);
                 armor.BasePowerUpAttributes.Add(powerUp);

